### DISCUSSION TITLE
Micrometer 1.13 migration path

### DISCUSSION THREAD

**Discussion by brunobat** [quarkusio/quarkus#42590] [Design Discussions]
This is a discussion to settle how the Micrometer 1.12 migration will be performed.
Given that the next LTS has been cut already this causes a problem for Keycloak. See https://github.com/quarkusio/quarkus/issues/42517

Initially we had plans to create the OTel Bridge for Micrometer and mark as deprecated the prometheus registry... See this message: https://groups.google.com/g/quarkus-dev/c/y5-ojIVsa_M/m/4wquJi4bBQAJ

The current plan is:
- We will mark the current extension deprecated.
- Implement the Micromenter to OTel bridge.
- Implement the prometheus registry 1.13.x on Quakiverse.

Please note that we want to align telemetry output around the OTel exporters. 
Even in the bridge is not ready, there is already a micrometer otlp registry on quarkiverse.

Is this plan ok for everyone?

CC @tqvarnst, @maxandersen @ebullient @vmuzikar

**Comment by brunobat** [comment]
Realistically, we might need to do the upgrade to 1.13 and backport to the LTS.

**Comment by vmuzikar** [comment]
@brunobat Thank you for opening this discussion.

If I understand correctly, Quarkus is moving away from Micrometer to OTel in any case, it's just a matter of when. To avoid some of the breaking changes, there will be a bridge/adapter to expose Micrometer API to applications that implement custom metrics so that they could keep using the Micrometer API but it would be served by OTel. However, with the switch to OTel there would still be breaking changes:
* No [default metrics as Micrometer had](https://quarkus.io/guides/telemetry-micrometer#automatically-generated-metrics).
* No OOTB support for OpenMetrics/Prometheus format using an [OTel exporter](https://quarkus.io/guides/opentelemetry#exporters). Micrometer [provided support for OpenMetrics format](https://quarkus.io/guides/telemetry-micrometer#micrometer-and-monitoring-system-extensions).
Please correct me if I'm wrong, I might have misunderstood the proposal.

This is IMHO a bit too big of a breaking change to do in a Quarkus minor release. I understand Micrometer would still be present in Quarkiverse but that means it's not supported e.g. in RHBQ.

Therefore I believe it would be good to get Micrometer 1.13 to the LTS. Problem might be the timing as the current 1.12 [reaches EOL in a few months](https://micrometer.io/support/).

CC @mabartos @ahus1

**Reply by ebullient** [reply]
The reason we didn't move from 1.12 to 1.13 is because it (in itself) introduces some breaking changes in terms of how it handles prometheus dependencies.

"moving away from Micrometer to OTel" -- don't misinterpret this. To my knowledge, the micrometer API isn't going anywhere.

**Reply by brunobat** [reply]
The story is a bit different. Quarkus will stick with Micrometer and all produced metrics will still be Micrometer, however, instead of exporting data with the prometheus registry, Quarkus will do it through OTel, with a bridge to be implemented.

We still plan to provide the prometheus registry but only on Quarkiverse...

Saying this, we still need to provide the current prometheus exporter on this LTS and mark it as deprecated, at some point.

**Reply by vmuzikar** [reply]
Thanks for the clarification. Do I understand correctly then that only the Prometheus registry will be deprecated and eventually removed? Micrometer as such is staying?

Do you plan to provide an OTel OpenMetrics exporter as a replacement for the Prometheus registry?

**Reply by brunobat** [reply]
correct

**Reply by ahus1** [reply]
I know this a late comment. I'm happy to read that Quarkus will continue to provide a metrics endpoint that exposes the OpenMetrics format. 

To add more context to the OTEL exporter vs. the OpenMetrics endpoint: When polling the metrics via a HTTP endpoint, this IMHO gives advantages compared to an OTEL exporter: 

* When pulling metrics from a REST endpoint, Prometheus can add context information to the metrics like the namespace, pod information, annotations on the pod, etc. All that information can be centrally configured in Prometheus. If the Pod or the OTEL exporter pushes that information, it would first need to acquire all that information from somewhere else, usually it needs to be passed into the Pods via a configuration. We're currently seeing this in the Keycloak project where we need to do that to get additional context for the traces. In summary, this is error prone, per and application. 
* When pushing any context information from a Pod, the information can be wrong - either by accident (a misconfiguration), or on purpose (considering there is a malicious actor). When using the Prometheus "pull", all that context information is added by Prometheus, and can therefore be trusted. 
* When polling metrics via Prometheus, this automatically provides the "up" metrics that Prometheus was able to talk to the Pod. This is an active information that can be used for alerting as well. Once the polling fails, all other metrics are outdated. With OTEL metrics and a push, all this information is missing and the listener can only assume a timeout. 

As a summary, I'm hoping that the OpenMetrics endpoint will stay, as it provides advantages over the OTEL exporter. 

If you know any remedies to the things described above in the OTEL space, I'm all ears.

**Reply by brunobat** [reply]
@trask, is there an answer to these concerns? Especially to the last point, which seems not to have any mitigation...

**Reply by trask** [reply]
hi @ahus1! OpenTelemetry has both push-based OTLP metric exporters and pull-based Prometheus (OpenMetrics) exporters

there's lots of details [here](https://opentelemetry.io/docs/specs/otel/compatibility/prometheus_and_openmetrics/), [here](https://opentelemetry.io/docs/specs/otel/metrics/sdk_exporters/prometheus/), and [here](https://opentelemetry.io/docs/languages/java/exporters/#prometheus)

**Reply by brunobat** [reply]
thanks @trask

**Comment by brunobat** [comment]
What do you guys think of using the `micrometer-registry-prometheus-simpleclient` (marked as deprecated) available on 1.13 and compatible with Prometheus Java client 0.x ?
This would minimize changes.

**Comment by brunobat** [comment]
Created this PR: https://github.com/quarkusio/quarkus/pull/42661

**Comment by maxandersen** [comment]
coming back from pto ..having one main question: When are those steps planned for? 

to happen in LTS, after LTS, backported?

**Reply by cescoffier** [reply]
Before LTS  (need to be included in the LTS).

**Reply by brunobat** [reply]
Will need backporting.

**Reply by geoand** [reply]
Sounds like it should not be a problem, no?

**Reply by brunobat** [reply]
We need ppl to use try it out. I've reached to @ppalaga and @vmuzikar to have a look.

**Reply by ppalaga** [reply]
I have found a way how to solve the problems we had in QCXF and I am hereby acking bacporting Micrometer 1.13 to Quarkus 3.15. See https://github.com/quarkusio/quarkus/issues/42800#issuecomment-2326417803

**Comment by brunobat** [comment]
Upgrade done and backported to 3.14.
